### PR TITLE
Bugfix update connect vraserver function

### PR DESCRIPTION
**Pull request by Stevio54**
I have a new update to the connect commandlet. I have removed the need for -UserAttribute, simply by handling a 400 error and trying a second time by tacking that @DOMAIN portion of the username back on. This makes this issue transparent to the user. 

I have also added a step where after using the CSP login endpoint, I then pass its refresh token to the IaaS Login API (documented in 8.3 release notes as the desired process). This is all tested against 8.2 and 8.3 versions of vRA.

### CODE REVIEW

File: src/Functions/Public/Connect-vRAServer.ps1 [hunk 1]
```diff
@@ -221,16 +214,49 @@
...
                 }
 
-                $Response = Invoke-RestMethod @Params
+                # Here we try with the first set of params, if it fails, it may be due to another configuration
+                # so we try again with the alternative option if available
+                try {
+                    $Response = Invoke-RestMethod @Params
+                } catch [Exception]{
+                    if ($_.Exception.Response.StatusCode -eq "BadRequest" -and $null -ne $AlternateJson) {
+                        $Params = @{
```
Comment (chelnak) [inline src/Functions/Public/Connect-vRAServer.ps1:223]: I think the only thing you need to overwrite here is $Body... the Params splat is already in memory and can be reused from above.

File: src/Functions/Public/Connect-vRAServer.ps1 [hunk 2]
```diff
@@ -176,24 +162,31 @@
...
 
                     # --- Logging in with a domain
-                    if (@("SAM", "SAMAccountName").Contains($UserAttribute) -and $Username -match '@') {
+                    $URI = "https://$($Server)/csp/gateway/am/idp/auth/login?access_token"
+                    if ($Username -match '@') {
                         # Log in using the advanced authentication API
-                        $URI = "https://$($Server)/csp/gateway/am/idp/auth/login?access_token"
                         $User = $Username.split('@')[0]
                         $Domain = $Username.split('@')[1]
                         $JSON = @{
```
Comment (chelnak) [inline src/Functions/Public/Connect-vRAServer.ps1:170]: Given that `JSON` and `AlternativeJson` hastables have the same structure, was wondering if we could do something clever and reuse a single hashtable but overwrite properties `if` they need updating.

**Review by chelnak — COMMENTED** [review]
Looking good - left a few comments for discussion.

Removal of a param is also a breaking change so unless anybody disagrees this will be a Major version increment!

File: src/Functions/Public/Connect-vRAServer.ps1 [hunk 3]
```diff
@@ -221,16 +214,49 @@
 
                 }
 
-                $Response = Invoke-RestMethod @Params
+                # Here we try with the first set of params, if it fails, it may be due to another configuration
+                # so we try again with the alternative option if available
+                try {
+                    $Response = Invoke-RestMethod @Params
+                } catch [Exception]{
```
[the remark this answers is not in the document]
Comment (Stevio54) [reply src/Functions/Public/Connect-vRAServer.ps1:221]: Let me look into it

File: src/Functions/Public/Connect-vRAServer.ps1 (hunk 1 above)
Comment (Stevio54) [reply src/Functions/Public/Connect-vRAServer.ps1:223]: Will make the change

File: src/Functions/Public/Connect-vRAServer.ps1 (hunk 2 above)
Comment (Stevio54) [reply src/Functions/Public/Connect-vRAServer.ps1]: updated the code, take a look, should show $RawBody in place of $JSON variables, then I convert on the call to Invoke

File: src/Functions/Public/Connect-vRAServer.ps1 (hunk 1 above)
Comment (Stevio54) [reply src/Functions/Public/Connect-vRAServer.ps1]: Did this now, please take a look

File: src/Functions/Public/Connect-vRAServer.ps1 (hunk 3 above)
[the remark this answers is not in the document]
Comment (Stevio54) [reply src/Functions/Public/Connect-vRAServer.ps1]: System.Net.WebException is the type, I am only catching this now, the main wrapping Try Catch will catch other errors.

**Comment by Stevio54** [comment]
I am closing this pull request. Submitting a new PR off another branch with just the bugfix.

**Pull request closed without merging**